pkg/cli/internal/cmd/preflight/checkers.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2019 The KubeCarrier Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package preflight
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gernest/wow"
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	versionutil "k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kubermatic/kubecarrier/pkg/cli/internal/spinner"
+	"github.com/kubermatic/kubecarrier/pkg/internal/util"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(appsv1.AddToScheme(scheme))
+	utilruntime.Must(corev1.AddToScheme(scheme))
+}
+
+const (
+	firstSupportedKubernetesVersion = "v1.16.0"
+
+	certManagerNamespaceName            = "cert-manager"
+	certManagerCAInjectorDeploymentName = "cert-manager-cainjector"
+	certManagerWebhookDeploymentName    = "cert-manager-webhook"
+	certManagerDeploymentName           = "cert-manager"
```

Comment:
We are not going to hard-code how the cert-manager should be deployed in the cluster. That is the responsibility of the system administration setting up the cluster and maintaining it. 

What this check should do is the following --> verify the cert-manager CRDs are present & installed, and at the right version. That's it; everything else is a too strong coupling between how the cert-manager is deployed and kubecarrier.

Reply:
but it can be the case that cert-manager CRDs are installed but the deployments are not ready, which means even if the preflight checks are passed, the `kubecarrier install setup` will still fail.
Also for checking the CRDs, I also need to hard-code the CRD names, right?

Reply:
what do you think @thetechnick @aborilov ?

Reply:
CRD names are cluster scoped and more stable. Furthermore, we explicitly depend on that CRD names when creating cert-manager objects in the kubecarrier itself. Thus it's a direct dependency that those CRDs exist.

Reply:
> but it can be the case that cert-manager CRDs are installed but the deployments are not ready, which means even if the preflight checks are passed

that is true. But with your approach, you're enforcing exact deployment name & namespace. You could still do this, but mark it only as a warning if they are not there. Something like. 

`[WARN] cert-manager deployment not found in cert-manager namespace. The kubecarrier requires working cert-manager deployment. You're responsible for satisfying this condition.`

Reply:
`cert-manger` namespace is the default namespace to install cert-manager from cert-manger official document: https://cert-manager.io/docs/installation/kubernetes/#installing-with-regular-manifests
They also mentioned the following notes:

Note: By default, cert-manager will be installed into the cert-manager namespace. It is possible to run cert-manager in a different namespace, although you will need to make modifications to the deployment manifests.

I am not really sure this will happen a lot, since I don't really know what is the use case for some people to deploy cert-manger in some other namesapce, and introduce extra complexity to edit the manifests.

What do you think @thetechnick ?

Reply:
I also think that checking CRDs is enough.
If we need more detailed error handling, the KubeCarrier Operator should report more status details, like: `CertificatesReady` or similar. That way we can show meaningful error messages to the enduser, without having us to rely on too much hardcoded information.